### DISCUSSION TITLE
brew doctor should not make changes

### DISCUSSION THREAD

**Discussion by machale** [Homebrew/discussions#4248] [Everyday usage] [answered]
### Output of `brew config`

```shell
% brew config
HOMEBREW_VERSION: 4.0.3
ORIGIN: https://github.com/Homebrew/brew
HEAD: ba6d87eed9ce31f42fa2b33a4e6067d9cf17f13f
Last commit: 9 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 45be88f3e5229a52c55589744a7ce583382566b8
Core tap last commit: 2 days ago
Core tap branch: master
Core tap JSON: 21 Feb 02:05 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_ANALYTICS: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit icelake
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
```

### Output of `brew doctor`

```shell
% brew doctor
==> Downloading https://formulae.brew.sh/api/formula.json
######################################################################## 100.0%
Your system is ready to brew.
```

### Description of issue

`brew doctor` generated the output above, one time. I can not repeat this, but I always thought `brew doctor` should make no changes to my system.

My main concern is `brew doctor` making changes, but as another data point, `brew update` did something similar - also only once.

```
% brew update
==> Downloading https://formulae.brew.sh/api/cask.json

Already up-to-date.
```

This occurred when I ran `brew doctor` approximately US EST 9:00pm on Monday February 20. I recently (February 18) upgraded macOS from 13.2 to 13.2.1, but this was not the first time since then that I ran `brew doctor`, so I don't _suspect_ this is related to the macOS upgrade.

**Comment by machale** [comment]
Hmm. Right after submitting my post above, I tried `brew doctor` again and the behavior _did_ repeat.

```
% brew doctor
==> Downloading https://formulae.brew.sh/api/formula.json
######################################################################## 100.0%
Your system is ready to brew.
```

A subsequent `brew update` did not repeat the behavior of downloading cask.json. And an additional attempt with `brew doctor` also did not repeat the behavior of dowbloading formula.json.

As of now, I don't have a pattern where this does, or does not, repeat.

**Comment by ZhongRuoyu** [comment, marked as answer]
`brew doctor` performs a series of checks, and some of them require the list of formulae (e.g. checking if any formulae installed are deprecated). So when you run `brew doctor`, the list of formulae will be downloaded from the API. Other than that, no changes will be made to your system.

> as another data point, brew update did something similar - also only once.

`brew update` updates Homebrew itself, and also fetches the list of formulae from API.

> As of now, I don't have a pattern where this does, or does not, repeat.

By default, `brew` checks the API every 450 seconds (7.5 minutes; see [`HOMEBREW_API_AUTO_UPDATE_SECS`](https://docs.brew.sh/Manpage#environment)). In other words, it won't check the API again if it already did so less than 7.5 minutes ago.

**Reply by machale** [reply]
See also the reply below from @ZhongRuoyu and their link to the 4.0.0 release notes. GitHub doesn’t show an option to mark that reply as the answer, so I’m marking this one instead.

**Comment by machale** [comment]
Ah, interesting that `brew doctor` needs to download info like this. For `brew update` it makes sense to me. I just always figured `brew doctor` only checked local stuff - but I admit that was an assumption on my part.

Is it a recent change that `brew doctor` and `brew update` print a line about the download? Otherwise it's surprising I've never encountered these printed lines in _many_ years of using Homebrew. (I probably average running `brew doctor` 2 or 3 times a week.)

**Reply by ZhongRuoyu** [reply]
> Is it a recent change that `brew doctor` and `brew update` print a line about the download?

Yes. We recently switched to using a JSON API for installation (and querying) of formulae. You may read the [4.0.0 release notes](https://brew.sh/2023/02/16/homebrew-4.0.0/) for more details.
